### PR TITLE
feat: Add support for `bool` (valueType) and `Boolean` (valueContent)

### PR DESCRIPTION
**Pull request by CJ42**
### What kind of change does this PR introduce (bug fix, feature, docs update, ...)?

⭐ Feature

### What is the current behaviour (you can also link to an open issue here)?

It is not possible to use erc725.js library to:
- encode/decode `bool` or an array of `bool[]` as value types.
- decode data keys that contain `Boolean` in the ERC725Y JSON Schema provided.

### What is the new behaviour (if this is a feature change)?

- add support for encoding/decoding `bool` and `bool[]` as value type.
- add support for decoding data to be interpreted as `Boolean`

### Other information:

Supersedes + closes #212 
closes issue #202

### CODE REVIEW

File: test/mockSchema.ts [hunk 1]
```diff
@@ -789,4 +789,30 @@ export const mockSchema: (ERC725JSONSchema & {
...
+  //     keyType: 'Singleton',
+  //     valueContent: 'Boolean',
+  //     valueType: 'bool',
+  //     // Test data
+  //     returnRawData: abiCoder.encodeParameter('bool', false),
+  //     returnRawDataArray: abiCoder.encodeParameter('bool[]', [false]),
+  //     returnGraphData: false,
+  //     expectedResult: false,
+  //   },
 ];
```
Comment (Hugoo) [inline test/mockSchema.ts:818]: why this commented code?
These comments are super legacy - i don't see why we would need to add something in this part as a comment?

**Comment by Hugoo** [comment]
@CJ42 the standard mentions `boolean`, not `bool`:
https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-2-ERC725YJSONSchema.md#valuetype

image

Shall we:

1. Update the standard to **add** `bool` AND add `boolean` support in this PR?
2. Update the standard to **replace** `boolean` to `bool`?
3. Change this PR to support `boolean` instead of `bool` ?

I think 3. is the way to go?

**Comment by CJ42** [comment]
@Hugoo that's a very good point to have spotted! 👌 
I can change it to `boolean` for now.

However, we should either:
- go with option 2) in the future
- allow both `bool` and `boolean`

So in the LSP2 standard, change `boolean` by `bool`. Simply because the `valueType` is for encoding/decoding in the smart contract, according to the smart language (Solidity in our case). 

image

In Solidity, the type is `bool` ✅ (not `boolean` ❌ ).
However☝️ `boolean` is quite common. So maybe we should allow both?

**Comment by CJ42** [comment]
@Hugoo we changed `boolean` by `bool` in LSP2: https://github.com/lukso-network/LIPs/pulls
In the future we should allow both to write either `boolean` or `bool`. `boolean` would fallback to encoding/decoding as `bool`. Opened a feature request: https://github.com/ERC725Alliance/erc725.js/issues/268`

**Comment by CJ42** [comment]
Implemented `boolean` as well. Closes #268

**Review by samuel-videau — APPROVED** [review]

File: src/lib/encoder.ts [hunk 2]
```diff
@@ -170,6 +178,10 @@ const valueTypeEncodingMap = {
     encode: (value: string) => abiCoder.encodeParameter('bytes', value),
     decode: (value: string) => abiCoder.decodeParameter('bytes', value),
   },
+  'bool[]': {
+    encode: (value: boolean) => abiCoder.encodeParameter('bool[]', value),
+    decode: (value: string) => abiCoder.decodeParameter('bool[]', value),
+  },
```
Comment (Hugoo) [inline src/lib/encoder.ts:184]: You added `boolean` but I think you would need to also add it here with `boolean[]`

File: src/types/ERC725JSONSchema.ts [hunk 3]
```diff
@@ -31,6 +32,7 @@ export type ERC725JSONSchemaValueType =
   | 'bytes32[]'
   | 'bytes4[]'
   | 'bytes[]'
+  | 'boolean'
```
Comment (Hugoo) [inline src/types/ERC725JSONSchema.ts:35]: so after the discussions, this should be `bool` not `boolean` ?

**Review by Hugoo — CHANGES_REQUESTED** [review]
i think there are still a few missing things

File: src/types/ERC725JSONSchema.ts (hunk 3 above)
Comment (Hugoo) [reply src/types/ERC725JSONSchema.ts:35]: or both as you said you want to support both

File: src/lib/encoder.ts (hunk 2 above)
Comment (CJ42) [reply src/lib/encoder.ts:184]: @Hugoo good catch.
I have added the return type `boolean[]` and some more tests.

File: src/types/ERC725JSONSchema.ts (hunk 3 above)
Comment (CJ42) [reply src/types/ERC725JSONSchema.ts]: @Hugoo yes should support both in the library: `bool` and `boolean`. Added the types

**Review by Hugoo — APPROVED** [review]

**Review by Hugoo — APPROVED** [review]

**Pull request merged**